### PR TITLE
Reuse kernel even if notebook was closed

### LINKED ISSUE #2238: Reuse running notebook kernel if notebook is closed
Say a user opens a notebook, runs some code, closes the notebook. The kernel will still be running. Now the user re-opens the notebook, and we should make sure that we re-use the already running kernel, like JupyterLab does.

### PR DESCRIPTION
**Pull request by davidanthoff**
Fixes #2238.

@DonJayamanne, does this look ok to you? It feels a bit ad-hoc, but I guess this is how we would do this kind of thing?

### CODE REVIEW

File: src/notebook/notebookFeature.ts [hunk 1]
```diff
@@ -57,15 +57,31 @@ export class JuliaNotebookFeature {
...
-                if (this.kernels.get(notebook) === kernel) {
-                    this.kernels.delete(notebook)
+            // Check whether there is still a running kernel for a closed notebook with the same Uri
+            let foundExistingKernel = false
+            for (const [k, v] of this.kernels) {
+                if (k.isClosed && (k.uri.toString() === notebook.uri.toString())) {
+                    foundExistingKernel = true
+                    v.notebook = notebook
+
+                    this.kernels.delete(k)
```
Comment (DonJayamanne) [inline src/notebook/notebookFeature.ts:68]: Does this do anything, the key in the map is `notebook`, hence this line is a noop.

File: src/notebook/notebookFeature.ts [hunk 2]
```diff
@@ -57,15 +57,31 @@ export class JuliaNotebookFeature {
...
-            await this.workspaceFeature.addNotebookKernel(kernel)
-            this.kernels.set(notebook, kernel)
 
-            kernel.onStopped(e => {
-                if (this.kernels.get(notebook) === kernel) {
-                    this.kernels.delete(notebook)
+            // Check whether there is still a running kernel for a closed notebook with the same Uri
+            let foundExistingKernel = false
+            for (const [k, v] of this.kernels) {
+                if (k.isClosed && (k.uri.toString() === notebook.uri.toString())) {
```
Comment (DonJayamanne) [inline src/notebook/notebookFeature.ts:64]: Shouldn't this be `if (!k.isClosed`

File: src/notebook/notebookFeature.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/notebook/notebookFeature.ts:68]: Rest seem ok.

**Review by DonJayamanne — APPROVED** [review]

File: src/notebook/notebookFeature.ts (hunk 1 above)
Comment (davidanthoff) [reply src/notebook/notebookFeature.ts:68]: The naming here is confusing, the `k` is the key, so it is the previous (closed) notebook, so this line removes that.

**Pull request merged**